desc/compute/_basis_vectors.py

```diff
@@ -3436,8 +3441,11 @@ def _gradpsi(params, transforms, profiles, data, **kwargs):
 def _n_rho(params, transforms, profiles, data, **kwargs):
     # equal to e^rho / |e^rho| but works correctly for surfaces as well that don't have
     # contravariant basis defined
-    data["n_rho"] = (
-        cross(data["e_theta"], data["e_zeta"]) / data["|e_theta x e_zeta|"][:, None]
+    data["n_rho"] = transforms["grid"].replace_at_axis(
+        (cross(data["e_theta"], data["e_zeta"]).T / data["|e_theta x e_zeta|"]).T,
+        lambda: (
+            cross(data["e_theta_r"], data["e_zeta"]).T / data["|e_theta x e_zeta|_r"]
```

Comment:
is this correct? I would think that the rho normal would be undefined at the axis?

Reply:
I'll change it to return `jnp.nan`. Interestingly, this is the correct multivariable limit of $n_{\rho}$ at the magnetic axis; all sequences converge to this vector. The problem is that once at the limit point, this vector becomes (finite) multi valued. So the geometric definition of $n_{\rho}$ is no longer equivalent to this formula. (the parameterization fails).

Reply:
@f0uriest What I did originally was correct. We should return `cross(data["e_theta_r"], data["e_zeta"]).T / data["|e_theta x e_zeta|_r"]`. This is the correct limit. The multivalued-ness is a feature not a bug. Below is my explanation.

The geometric interpretation of $n_{\rho}$ as a **vector** is of course undefined at the magnetic axis because $n_{\rho}$ becomes multivalued. Many other quantities such as $\partial^{n \geq 1}\_{\rho} e_{\theta}$, $e^{\rho}$, and some of the metric tensor elements are also (finite) multivalued at the magnetic axis. In particular, the force balance error **vector** $\mathbf{F}$ is **not**  defined at the magnetic axis unless its radial component is zero (usually not the case) and the force balance error magnitude $\lVert \mathbf{F} \rVert$ is not defined unless its radial component is either zero or the multivalued inverse of $\lVert e^{\rho} \rVert$ (very unlikely) so that the single-valuedness can be recovered.

I can share my proofs, but this shouldn't be that surprising. Even though force is a physical quantity, there is no reason for it be representable by a unique vector at a geometric singularity. Think of the pressure force when compressing a balloon. It can represented as a single-valued vector everywhere except at the center of the balloon where it must simultaneously be a vector in every direction.

We have many quantities that are well-defined as scalars or vectors everywhere except the geometric singularity $\rho = 0$. At $\rho = 0$, they are no longer scalars or vectors, but their functional definition is still well-defined. We need to retain the information in these functional definitions to be able to correctly evaluate other quantities.

As an example of why this is useful, think of the single-valued map that transforms into multivalued one (i.e. a set) at a geometric singularity $\rho = 0$ that is the partial derivative with respect to $\rho$ of the surface area Jacobian of an elliptical torus with major radius $R$ and semi-major radius $\rho$ and semi-minor $b$. At $\rho = 0$, this map is
$$(\partial_{\rho} \lVert \mathbf{e}\_{\theta} \times \mathbf{e}\_{\zeta} \rVert)_{\rho = 0} \colon (\rho=0, \theta, \zeta) \mapsto -(b R \cos{\zeta} \cos \theta) \hat{x}  - (b R \sin \zeta \cos \theta) \hat{y}  - (R \sin \theta) \hat{z} \in \mathbb{R}^3$$

If we want to treat the derivative of the surface area Jacobian as a scalar, then clearly it is undefined at the magnetic axis because it is multivalued there. But the derivative of the surface area itself: $\partial_{\rho} S = \\oint d\theta d\zeta (\partial_{\rho} \lVert \mathbf{e}\_{\theta} \times \mathbf{e}\_{\zeta} \rVert)$ or `S_r(r)` in the code is well defined as a scalar at the magnetic axis because the integration transforms the the multivalued map into a single valued one.

You can get error of up to 10% for $\partial_{\rho} S$ at the magnetic axis even for torus with elongation ~3 if you don't perform this eliptical integral using the functional definition of the integrand at the axis (so this is not just a technicality).

I don't see an issue with leaving $n_{\rho}$ as is instead of `jnp.nan` because it's going to correctly evaluate as multivalued anyway on a `LinearGrid` which has multiple theta nodes at a fixed location in $\rho$ and $\zeta$.

Reply:
I copied the wrong equation for the last part and will update later but the point still stands

Reply:
OK, I buy that $F$ may be multivalued, since it's not necessarily divergence free. But $B$ and $J$ should definitely be single valued, so just want to make sure that's enforced.

Reply:
B is single-valued; see the new test in test_axis_limits. I'll check if J is

Reply:
I don't see why $\mathbf{J}$ needs to be single-valued. It's essentially the velocity vector of a charged particle. I'm thinking of a charged particle gyrating around a field line; this circular motion is multivalued at the center of the circle. And both $\mathbf{J} \times \mathbf{B} = \nabla p$ and $\nabla \times \mathbf{B} = \mathbf{J}$ suggest to me that: $\mathbf{B}$ is single-valued implies $\mathbf{J}$ is multivalued

Reply:
This is interesting stuff, Kaya! I will just say some random things that may or may not help.

I think the existence of grad rho should depend upon the derivation of rho. What if rho is like the half minor radius/half minor radius of the last surface? Or the enclosed poloidal flux instead of the toroidal flux. 

I also think analyzing the existence of grad rho would be easier in a cylindrical/cartesian coordinate system. We can take a constant phi surface and evaluate the derivative of rho w.r.t R and Z. If we choose polar coordinates, it becomes problematic because theta is not defined on the axis. Theta only makes sense when we choose a point, move away from it, and rotate about it.

Shouldn't grad p be zero on the axis? I don't mean dp/drho grad rho but dp/dX grad X + dp/dY grad Y + dp/dZ grad Z just from the regularity of the pressure. If I plot p vs X or p vs R, I feel like it should be smooth on the axis without any kinks.
If so, J = constant x B and single-valued. If not, I still don't see what single-values B implies multivalued J.

Ideal MHD requires gyroaveraging particle orbits and "high" collisionality. By that, you average the gyration and assume that the ion-ion collisions dominate. High collisionality causes the ions and electrons to behave like a single fluid. However, the effects of particle gyration (gyroviscosity) are subdominant in ideal MHD. So the particle motion should be random, both perpendicular and parallel to a field line on a scale much smaller than the gyroradius.

Maybe these things are undefined because when we get that close to the axis, something in ideal MHD breaks down, and we need kinetic theory.

Reply:
> Shouldn't grad p be zero on the axis?

I forgot that $\nabla p = \partial_{\rho} p * e^{\rho}$ was zero on the axis because $\partial_{\rho} p$ is zero at axis. In that case $\mathbf{J} \times \mathbf{B} = \text{constant} = 0$ so I can't conclude $\mathbf{J}$ is multivalued. Before I was thinking $e^{\rho} \text{ multivalued} \implies \nabla p \text{ multivalued} \implies \mathbf{J} \times \mathbf{B} = \nabla p \text{ multivalued} \implies \mathbf{J} \text{ multivalued because } \mathbf{B} \text{ single-valued}$

(Though without additional constraints that may come from the physics, there are still an infinite number of vectors that are orthogonal to $\mathbf{B}$ and could satisfy $\mathbf{J} \times \mathbf{B} = 0$).

When I compute $\mathbf{J}$ at axis in this PR now there is some variation in $\theta$ of the order $10^{-9}$. I think this is just round off error, even though the variation in $\mathbf{B}$ is of the order $10^{-32}$.

Reply:
I agree that sounds like round-off error for $J$, and that makes sense that it is single-valued at the axis. $J$ is like the derivative of $B$ and also the magnitude of the current is much larger than the magnetic field in our units, so considering the relative errors I think this is an expected order of magnitude.

Reply:
I agree. Physical quantities such as $\mathbf{B}$ and $p$ should be analytic functions of the lab frame coordinates, and hence single valued everywhere. Likewise, their derivatives (like $\mathbf{J}$ and $\mathbf{F}$) would also have to be single valued. In your balloon example, the pressure at the center isn't pushing in every direction, the force at the center is zero, similar to how $dp/d\rho = 0$ on axis for physical profiles.

1e-9 seems reasonable for rounding error for J, the round of error in `curl(B)` (which is what is really calculated, then multiplied by mu0) would then be ~1e-16. I've found that even on "zero net current" cases where we fix the current profile to zero, the resulting current from taking the FSA ends up being ~1e-10 which tracks with what you're seeing.